Type error in `async` function causes spurious "type must be known in this context" error

The following code:

```rust
async fn weird() {
    1 = 2;
    
    let mut loop_count = 0;
    async {}.await
}

fn main() {}
```

produces the following errors:

```
error[E0070]: invalid left-hand side of assignment
 --> src/main.rs:2:7
  |
2 |     1 = 2;
  |     - ^
  |     |
  |     cannot assign to this expression

error[E0698]: type inside `async fn` body must be known in this context
 --> src/main.rs:4:9
  |
4 |     let mut loop_count = 0;
  |         ^^^^^^^^^^^^^^ cannot infer type for type `{integer}`
  |
note: the type is part of the `async fn` body because of this `await`
 --> src/main.rs:5:5
  |
5 |     async {}.await
  |     ^^^^^^^^^^^^^^

error: aborting due to 2 previous errors
```

The "type inside `async fn` body must be known in this context" error is spurious, and disappears if the  `1 = 2;` line is removed.

--- a/compiler/rustc_typeck/src/check/generator_interior.rs
+++ b/compiler/rustc_typeck/src/check/generator_interior.rs
@@ -89,19 +89,31 @@ impl<'a, 'tcx> InteriorVisitor<'a, 'tcx> {
             if let Some((unresolved_type, unresolved_type_span)) =
                 self.fcx.unresolved_type_vars(&ty)
             {
-                let note = format!(
-                    "the type is part of the {} because of this {}",
-                    self.kind, yield_data.source
-                );
-
                 // If unresolved type isn't a ty_var then unresolved_type_span is None
                 let span = self
                     .prev_unresolved_span
                     .unwrap_or_else(|| unresolved_type_span.unwrap_or(source_span));
-                self.fcx
-                    .need_type_info_err_in_generator(self.kind, span, unresolved_type)
-                    .span_note(yield_data.span, &*note)
-                    .emit();
+
+                // If we encounter an int/float variable, then inference fallback didn't
+                // finish due to some other error. Don't emit spurious additional errors.
+                if let ty::Infer(ty::InferTy::IntVar(_) | ty::InferTy::FloatVar(_)) =
+                    unresolved_type.kind()
+                {
+                    self.fcx
+                        .tcx
+                        .sess
+                        .delay_span_bug(span, &format!("Encountered var {:?}", unresolved_type));
+                } else {
+                    let note = format!(
+                        "the type is part of the {} because of this {}",
+                        self.kind, yield_data.source
+                    );
+
+                    self.fcx
+                        .need_type_info_err_in_generator(self.kind, span, unresolved_type)
+                        .span_note(yield_data.span, &*note)
+                        .emit();
+                }
             } else {
                 // Insert the type into the ordered set.
                 let scope_span = scope.map(|s| s.span(self.fcx.tcx, self.region_scope_tree));

--- a/src/test/ui/async-await/issue-73741-type-err.rs
+++ b/src/test/ui/async-await/issue-73741-type-err.rs
@@ -1,0 +1,14 @@
+// edition:2018
+//
+// Regression test for issue #73741
+// Ensures that we don't emit spurious errors when
+// a type error ocurrs in an `async fn`
+
+async fn weird() {
+    1 = 2; //~ ERROR invalid left-hand side
+
+    let mut loop_count = 0;
+    async {}.await
+}
+
+fn main() {}

--- a/src/test/ui/async-await/issue-73741-type-err.stderr
+++ b/src/test/ui/async-await/issue-73741-type-err.stderr
@@ -1,0 +1,11 @@
+error[E0070]: invalid left-hand side of assignment
+  --> $DIR/issue-73741-type-err.rs:8:7
+   |
+LL |     1 = 2;
+   |     - ^
+   |     |
+   |     cannot assign to this expression
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0070`.

